faiss_vector_cache.go

```diff
@@ -0,0 +1,228 @@
+		// 	this makes the average to be kept above the threshold value for a
+		// 	longer time and thereby the index to be resident in the cache
+		// 	for longer time.
+		vc.cache[fieldIDPlus1] = initCacheEntry(index, 0.4)
+	}
+}
+
+func (vc *vecIndexCache) isIndexCached(fieldID uint16) (*faiss.IndexImpl, bool) {
+	vc.m.RLock()
+	entry, present := vc.cache[fieldID]
+	vc.m.RUnlock()
+	if entry == nil {
+		return nil, false
+	}
+
+	rv := entry.index
+	return rv, present && (rv != nil)
+}
+
+func (vc *vecIndexCache) incHit(fieldIDPlus1 uint16) {
+	vc.m.RLock()
+	entry := vc.cache[fieldIDPlus1]
+	entry.incHit()
+	vc.m.RUnlock()
+}
+
+func (vc *vecIndexCache) addRef(fieldIDPlus1 uint16) {
+	vc.m.RLock()
+	entry := vc.cache[fieldIDPlus1]
+	entry.addRef()
+	vc.m.RUnlock()
+}
+
+func (vc *vecIndexCache) decRef(fieldIDPlus1 uint16) {
+	vc.m.RLock()
+	entry := vc.cache[fieldIDPlus1]
+	entry.decRef()
+	vc.m.RUnlock()
+}
+
+func (vc *vecIndexCache) refresh() bool {
+	vc.m.Lock()
+	cache := vc.cache
+
+	// for every field reconcile the average with the current sample values
+	for fieldIDPlus1, entry := range cache {
+		sample := atomic.LoadUint64(&entry.tracker.sample)
+		entry.tracker.add(sample)
+
+		refCount := atomic.LoadInt64(&entry.refs)
+		// the comparison threshold as of now is (1 - a). mathematically it
+		// means that there is only 1 query per second on average as per history.
+		// and in the current second, there were no queries performed against
+		// this index.
+		if entry.tracker.avg <= (1-entry.tracker.alpha) && refCount <= 0 {
+			atomic.StoreUint64(&entry.tracker.sample, 0)
+			entry.closeIndex()
+			delete(vc.cache, fieldIDPlus1)
+			continue
+		}
+		atomic.StoreUint64(&entry.tracker.sample, 0)
+	}
+
+	rv := len(vc.cache) == 0
+	vc.m.Unlock()
+	return rv
+}
+
+func (vc *vecIndexCache) monitor() {
+	ticker := time.NewTicker(1 * time.Second)
+	for {
+		select {
+		case <-vc.closeCh:
+			return
+		case <-ticker.C:
+			exit := vc.refresh()
+			if exit {
+				// no entries to be monitored, exit
+				return
+			}
+		}
+	}
+}
+
+func (e *ewma) add(val uint64) {
+	if e.avg == 0.0 {
+		e.avg = float64(val)
+	} else {
+		// the exponentially weighted moving average
+		// X(t) = a.v + (1 - a).X(t-1)
+		e.avg = e.alpha*float64(val) + (1-e.alpha)*e.avg
+	}
+}
+
+func initCacheEntry(index *faiss.IndexImpl, alpha float64) *cacheEntry {
+	vc := &cacheEntry{
+		index:   index,
+		tracker: &ewma{},
+	}
+	vc.tracker.alpha = alpha
+
+	atomic.StoreUint64(&vc.tracker.sample, 1)
+	return vc
+}
+
+func (vc *cacheEntry) incHit() {
+	// every access to the cache entry is accumulated as part of a sample
+	// which will be used to calculate the average in the next cycle of average
+	// computation
+	atomic.AddUint64(&vc.tracker.sample, 1)
+}
+
+func (vc *cacheEntry) addRef() {
+	atomic.AddInt64(&vc.refs, 1)
+}
+
+func (vc *cacheEntry) decRef() {
+	atomic.AddInt64(&vc.refs, -1)
+}
+
+func (vc *cacheEntry) closeIndex() {
```

Comment:
Rename this to closeIndexLOCKED().

Reply:
And it's contents need to be non blocking - because vc.Index.Close() could be long. So -

```
go func() {
    vc.index.Close()
    vc.index = nil
}()
```

Reply:
Maybe the function name doesn't need changing :)